src/services/ftp/util.rs

```diff
@@ -0,0 +1,64 @@
+use futures::AsyncRead;
+
+use crate::BytesReader;
+use std::io::Result;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use suppaftp::FtpStream;
+use suppaftp::Status;
+use tokio::sync::Mutex;
+use tokio::task;
+
+/// Wrapper for ftp data stream and command stream.
+pub struct FtpReader {
+    reader: BytesReader,
+    client: Arc<Mutex<FtpStream>>,
+}
+
+impl FtpReader {
+    /// Create an instance of FtpReader.
+    pub fn new(r: BytesReader, c: FtpStream) -> Self {
+        Self {
+            reader: r,
+            client: Arc::new(Mutex::new(c)),
+        }
+    }
+}
+
+impl AsyncRead for FtpReader {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize>> {
+        let result = Pin::new(&mut self.reader).poll_read(cx, buf);
+
+        // Drop data stream and close command stream when hit Error or EOF.
+        match result {
+            Poll::Ready(Err(_)) | Poll::Ready(Ok(0)) => {
+                let c = self.client.clone();
+
+                drop(self.reader.as_mut());
+
+                task::spawn(async move {
```

Comment:
Please don't spawn a new task. Instead, write our future state machine like:

https://github.com/datafuselabs/opendal/blob/75eb57e15ce491b87e536dbcfc944f02abdf43f1/src/io_util/into_reader.rs#L64-L67

In general, we should do the following thing when we meet an error or reading done:

- Build a new `BoxFuture`
- Switch state so that we can poll it
- Exit when we get a result

Reply:
`IntoReader` is not a good example, please take `DirStream` as a look:

https://github.com/datafuselabs/opendal/blob/75eb57e15ce491b87e536dbcfc944f02abdf43f1/src/services/s3/dir_stream.rs#L79-L103

Reply:
Thanks! I made change to the `poll_read()` and It should work as planned. My understanding of "Async Rust" has increased every day:smiling_face_with_three_hearts:.

Reply:
> My understanding of "Async Rust" has increased every day since I contribute to this project .

It's our pleasure!